src/rewrite_rnn.cpp

```diff
@@ -1099,6 +1140,146 @@ std::vector<operation> rewrite_rnn::lstm_actv_funcs(instruction_ref ins) const
     }
 }
 
+bool rewrite_rnn::is_variable_seq_lens(const program& prog, instruction_ref seq_lens) const
+{
+    bool is_var_lens = false;
+    if(seq_lens != prog.end())
+    {
+        if(seq_lens->can_eval())
+        {
+            auto arg_lens = seq_lens->eval();
+            std::vector<int64_t> vec_lens;
+            arg_lens.visit([&](auto l) { vec_lens.assign(l.begin(), l.end()); });
+            int64_t l = 0;
+            if(!vec_lens.empty())
+            {
+                l = vec_lens[0];
+            }
+            if(!std::all_of(vec_lens.begin(), vec_lens.end(), [&](auto v) { return v == l; }))
+            {
+                is_var_lens = true;
+            }
+        }
+        else
+        {
+            is_var_lens = true;
+        }
+    }
+
+    return is_var_lens;
+}
+
+std::size_t
+rewrite_rnn::get_seq_len(const program& prog, instruction_ref input, instruction_ref seq_lens) const
+{
+    bool is_var_lens = is_variable_seq_lens(prog, seq_lens);
+    auto input_shape = input->get_shape();
+    auto length      = input_shape.lens()[0];
+    if(!is_var_lens and seq_lens != prog.end())
+    {
+        auto arg_len = seq_lens->eval();
+        std::vector<std::size_t> vec_lens;
+        arg_len.visit([&](auto l) { vec_lens.assign(l.begin(), l.end()); });
+        length = vec_lens.empty() ? length : vec_lens[0];
+    }
+
+    return length;
+}
+
+instruction_ref rewrite_rnn::replace_last_hs_output(program& prog,
+                                                    instruction_ref ins,
+                                                    instruction_ref seq_lens,
+                                                    instruction_ref last_hs_output,
+                                                    op::rnn_direction dirct) const
+{
+    bool variable_seq_len = is_variable_seq_lens(prog, seq_lens);
+    instruction_ref result_ins{};
+    if(variable_seq_len)
+    {
+        result_ins = prog.insert_instruction(
+            std::next(ins), op::rnn_shift_output{"hidden_states", dirct}, ins, seq_lens);
+        prog.replace_instruction(ins, result_ins);
```

Comment:
This wont replace the last_hs_output placeholder.

Reply:
This is to replace the hs_outputs of the case of all sequences are the same(not the output we need) with the hs_outputs of the case of sequence are of variable length(final output)